server/submission/resolvers.js

```diff
@@ -68,6 +74,47 @@ const resolvers = {
 
       return newManuscript
     },
+    async finishSubmission(_, { data }, ctx) {
+      const manuscript = await db.selectId(data.id)
+      db.checkPermission(manuscript, ctx.user)
+      const newManuscript = lodash.merge({}, manuscript, data)
+
+      const { errorManuscript } = Joi.validate(newManuscript, manuscriptSchema)
+      if (errorManuscript) {
+        throw new Error(errorManuscript)
+      }
+
+      const {
+        cosubmission,
+        cosubmissionTitle,
+        cosubmissionId,
+      } = newManuscript.submissionMeta
+      const { errorCosubmission } = Joi.validate(
+        {
+          cosubmission,
+          cosubmissionTitle,
+          cosubmissionId,
+        },
+        cosubmissionSchema,
+      )
+      if (errorCosubmission) {
+        throw new Error(errorCosubmission)
+      }
+
+      newManuscript.submissionMeta.stage = 'QA'
+      const newManuscriptDb = db.manuscriptGqlToDb(newManuscript, ctx.user)
+      await db.update(newManuscriptDb, data.id)
+
+      const mailData = {
+        from: config.get('mailer.from'),
+        to: newManuscript.submissionMeta.author.email,
+        subject: 'Congratulations! You submitted your manuscript!',
+        text: 'Your manuscript has been submitted',
+        html: '<p>Your manuscript has been submitted</p>',
+      }
+      Email.send(mailData)
```

Comment:
Right now if something throws in a resolver the error will be sent back to the client by apollo, and the frontend will display the error page with what error was generated. This happens in all resolvers (e.g. even when storing the data to the backend every 5s) so the user can get an error page without doing anything. Should this be treated in a separated issue?

Reply:
Also, regarding the retry strategy, that sounds good, but I think it should be done at pubsweet component level.

Reply:
hmmm.... that didn't happen for me... I got no error page ... have you tested it?

Reply:
This function is async so needs to be `await`ed.

Reply:
You are right, I tested this only with the query. For some reason the server/apollo doesn't log anything if you throw in a query resolver (e.g. `currentSubmission`) but it prints the error if you throw it in a mutation resolver. Also, the error page will only be displayed if the `currentSubmission` query fails but when a mutation fails nothing is displayed (this can be changed).

So do we do this try-catch only for the e-mail sending component in this mutation? There are other operations in different queries/mutations that could throw (e.g. db operations) and nothing is done in that regards.

If yes, what should be the expected behavior? If the e-mail is not sent/failed should we show some error to the user? Do we use the error page for that?

I suggest that for now we just catch errors from operations that can throw and log something on the backend. This should be done as part of this issue for `finishSubmission` mutation and part of another issue for other queries/mutations. Some errors like failing to store data to the database should be shown to the user in a nice format. Also, successfully storing the manuscript to the database at this stage should show success to the user somehow (do we have wireframes for this?). If the e-mail fails I won't show something to the user since this is just a confirmation e-mail that their manuscript has been received, but I would log something on the backend.

Should I work out some retry strategy and do a MR to pubsweet for their component? (this can take a while) or do I hack something in this PR for the moment?

Reply:
@tamlyn I have my doubts about that. Like I said in my previous comment, if we don't show the result of the e-mail component to the user not sure if we should wait for it. It's fine to wait for it to be stored to the db (so we can show success/failure), but does it make sense to wait for it and show something like "failure to send your confirmation e-mail"?

Reply:
Sure, you don't have to await it but you absolutely do have to handle the error some way. The way you have it here will lead to an unhandled promise rejection and this will crash the server in future versions of node.